core/src/main/java/org/apache/iceberg/UpdateRequirement.java

```diff
@@ -0,0 +1,371 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.List;
+import java.util.Set;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+
+/** Represents a requirement for a {@link MetadataUpdate} */
+public interface UpdateRequirement {
+  void validate(TableMetadata base);
+
+  static Builder builderForCreate() {
+    return new Builder(null, false).requireCreate();
+  }
+
+  static Builder builderForReplace(TableMetadata base) {
+    Preconditions.checkNotNull(base, "Cannot create a builder from table metadata: null");
+    return new Builder(base, true).requireTableUUID(base.uuid());
+  }
+
+  static Builder builderFor(TableMetadata base) {
+    Preconditions.checkNotNull(base, "Cannot create a builder from table metadata: null");
+    return new Builder(base, false).requireTableUUID(base.uuid());
+  }
+
+  class AssertTableDoesNotExist implements UpdateRequirement {
+    public AssertTableDoesNotExist() {}
```

Comment:
Why is this public? It wasn't public in the original version.

Reply:
See my other comment. I think we should remove the builder and keep the public constructors to match `MetadataUpdate`.

Reply:
tests in the rest package are using some of the constructors, hence I made those public. I'll remove the builder and rework the API to match `MetadataUpdate`